### PR TITLE
Discussion Settings: Interface Layer

### LINKED ISSUE #4382: Discussion Settings
##### Defaults for New Posts::
- [x] Allow Comments
- [x] Send Pingbacks
- [x] Receive Pingbacks

##### Comments:
- [x] Require name and e-mail
- [x] Require users to Sign In
- [x] Close after
- [x] Sort by
- [x] Threading
- [x] Paging
- [x] Automatically Approve
- [x] Links in Comments

##### More:
- [x] Moderation
- [x] Blacklist

--

#### Mockup: Top Level
main screen

#### Mockup: Close After
close-commenting

#### Mockup: Sort By
3 sort by

#### Mockup: Threading
4 threading

#### Mockup: Paging
paging

#### Mockup: Auto-approve
automatically-approve

#### Mockup: Links
links-in-comments

#### Mockup: Moderation
6 moderation

#### Mockup: Blacklist
7 blacklist

### PR DESCRIPTION
**Pull request by jleandroperez**
#### Description:
In this PR we implement the Interface Layer for the Blog's Discussion Settings. This is, for the time being, only available to REST endpoints.

Closes #4382.

#### Scenario 1: Self Hosted
1. Fresh install the app
2. Log into a self hosted blog
3. Open `My Sites` > `Settings`

Note that the *Discussion* row should be hidden.

#### Scenario 2: Dotcom Account
1. Log into your dotcom account.
2. Tap over `My Sites`, and pick any of your websites.
3. Press `Settings` > `Discussion`.

Verify that the values onscreen match with Calypso's values. Calypso settings should be available at the exact same path: `My Sites` > `Site X` > `Settings` > `Discussion`.

Please, try updating iOS values, and verify that they make their way to the web.

--

Needs Review: @mattmiklic + @diegoreymendez 
Thanks in advance gentleman!!

### CODE REVIEW

**Comment by mattmiklic** [comment]
Right now I'm just testing with a dotcom blog; I've requested a self-hosted sandbox so I can test that later.

- Most settings don't seem to save until after I've backed out of the Settings screen. We'll want to change this so that they save immediately (the same change we recently made on Android).
- After changing the "Sort by", "Threading", "Paging", or "Automatically approve" options, when I go back to the Discussion settings screen, the old value is still shown. If I go back to Settings, then tap Discussion again, the new value is shown (I think this is related to the previous point, because the save doesn't happen until I go back to the top level Settings screen)
- When adding an entry to the Hold for Moderation or Blacklist screen, tapping "Done" on the keyboard has no effect. If possible, we should save the value and go back to the list when the user taps Done.

**Comment by jleandroperez** [comment]
@mattmiklic Thanks *A LOT* for your feedback sir!. Everything has been addressed.
Minor note: we're not supporting self hosted. Only WordPress.com / Jetpack sites should work.

We'll get back to it as soon as the new API PR is merged into .org.

Thank you!

File: WordPress/Classes/ViewRelated/Cells/PickerTableViewCell.swift [hunk 1]
```diff
@@ -0,0 +1,133 @@
+import Foundation
+import WordPressShared
+
+
+/**
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Cells/PickerTableViewCell.swift]: The documentation format changed.  The one you're using here doesn't work very well with Xcode + Swift.  This is what comes up when I `alt` + click on this class' name.

screen shot 2015-12-23 at 5 53 00 pm

Here's some [additional info](swift documentation) on the new documentation formats.

Here's an example of how a method documented with the new format looks like:

screen shot 2015-12-23 at 5 57 13 pm

Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Cells/PickerTableViewCell.swift]: I could provide a better example with a method if you need it, let me know!

**Comment by jleandroperez** [comment]
@diegoreymendez thanks for the documentation tips!. All updated, ready for another look.

Thank you!

**Comment by diegoreymendez** [comment]
@jleandroperez - Just found a small issue... these are the steps:

1. Make sure that "Comment author must fill out name and e-mail" is checked.
2. Open the site settings in the mobile app.
3. Open the site settings in calypso.
4. Go back one level in the mobile app, without changing anything.
5. Uncheck "Comment author must fill out name and e-mail" in Calypso and save the settings.
6. Open the site settings in the mobile app again.

Expected behaviour: the site settings in the mobile app should reflect the latest changes.
Current behaviour: it doesn't.

**Comment by diegoreymendez** [comment]
Following up on my comment above: it seems like changes in Calypso are not properly refreshed in the app.

It "feels" to me like the discussion settings should be refreshed when the VC is opened to fix this problem.

**Comment by diegoreymendez** [comment]
An additional issue with this, is that when saving the discussion settings in WPiOS, all discussion settings are saved (not just the changed ones).  This should not be a big deal if the settings are refreshed when opening, but if they're not refreshed properly this can roll back several changes applied in Calypso and go unnoticed.

**Comment by jleandroperez** [comment]
@diegoreymendez thanks for spotting this one! I'll address that in #4615 (since this PR is huuuge!).

Thank you!

**Comment by diegoreymendez** [comment]
:shipit:

**Comment by jleandroperez** [comment]
@diegoreymendez *THANK YOU*!!!

**Pull request merged**